### PR TITLE
create endpoint POST /register in class UserController and implement …

### PR DESCRIPTION
**Pull request by ZetenyiEmese**
Megcsináltam a POST /register endpointot, és minden hozzátartozó cuccot, leteszteltem Postmannel (nekem ok), kérlek nézzétek át. Emese

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -22,4 +22,6 @@ dependencies {
     runtimeOnly 'mysql:mysql-connector-java'
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
     testImplementation 'org.springframework.security:spring-security-test'
+    compile 'io.jsonwebtoken:jjwt:0.6.0'
+
```
Comment (balazsbarni) [inline build.gradle]: please remove the empty line

File: src/main/java/com/greenfoxacademy/goddesstribesbackend/controllers/UserController.java [hunk 2]
```diff
@@ -17,4 +24,54 @@ public UserController(UserService userService, KingdomService kingdomService) {
...
+      return ResponseEntity.status(400).body(new ErrorMessage("Username and password are required."));
+    }
+    if (username == null || username.isEmpty()) {
+      return ResponseEntity.status(400).body(new ErrorMessage("Username is required."));
+    }
+    if (password == null || password.isEmpty()) {
+      return ResponseEntity.status(400).body(new ErrorMessage("Password is required."));
+    }
+    if (userService.checkUserByName(username)) {
+      return ResponseEntity.status(400).body(new ErrorMessage("Username is already taken."));
```
Comment (balazsbarni) [inline src/main/java/com/greenfoxacademy/goddesstribesbackend/controllers/UserController.java]: 409 status by APIdocs and would be nice to respond : "Zolika -as username taken", ore something like that

File: src/main/java/com/greenfoxacademy/goddesstribesbackend/controllers/UserController.java [hunk 3]
```diff
@@ -17,4 +24,54 @@ public UserController(UserService userService, KingdomService kingdomService) {
...
+    if (username == null || username.isEmpty()) {
+      return ResponseEntity.status(400).body(new ErrorMessage("Username is required."));
+    }
+    if (password == null || password.isEmpty()) {
+      return ResponseEntity.status(400).body(new ErrorMessage("Password is required."));
+    }
+
+    if (!userService.checkUserByNameAndPassword(username, password)){
+      return ResponseEntity.status(401).body(new ErrorMessage("Username or password is incorrect."));
+    } else {
```
Comment (balazsbarni) [inline src/main/java/com/greenfoxacademy/goddesstribesbackend/controllers/UserController.java]: else is needed?

File: src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTAuthenticationFilter.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import java.io.IOException;
+
+public class JWTAuthenticationFilter extends GenericFilterBean {
+  @Override
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain)
+          throws IOException, ServletException {
+    Authentication authentication = JWTUtility.getAuthentication((HttpServletRequest) request);
+    SecurityContextHolder.getContext().setAuthentication(authentication);
+    filterChain.doFilter(request, response);
+
```
Comment (balazsbarni) [inline src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTAuthenticationFilter.java]: please remove the empty line

File: src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTUtility.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            .setSubject(username)
+            .signWith(SignatureAlgorithm.HS256, SecurityCostants.SECRET)
+            .compact();
+    return jwtToken;
+  }
+
+  public static Authentication getAuthentication(HttpServletRequest request) {
+    String token = request.getHeader(SecurityCostants.HEADER_STRING);
+    String username;
+
```
Comment (balazsbarni) [inline src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTUtility.java]: not needed

File: src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTUtility.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    if (token != null) {
+      try {
+        username = Jwts.parser()
+                .setSigningKey(SecurityCostants.SECRET)
+                .parseClaimsJws(token.replace(SecurityCostants.TOKEN_PREFIX, ""))
+                .getBody()
+                .getSubject();
+        return new UsernamePasswordAuthenticationToken(username, null, emptyList());
+
```
Comment (balazsbarni) [inline src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTUtility.java]: not needed

File: src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTUtility.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      } catch (SignatureException e) {
+        return null;
+      }
+    }
+    return null;
+  }
+
+}
```
Comment (balazsbarni) [inline src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTUtility.java]: empty line after the last line

**Review by balazsbarni — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 1 above)
Comment (ZetenyiEmese) [reply build.gradle]: Done.

File: src/main/java/com/greenfoxacademy/goddesstribesbackend/controllers/UserController.java (hunk 2 above)
Comment (ZetenyiEmese) [reply src/main/java/com/greenfoxacademy/goddesstribesbackend/controllers/UserController.java]: Done

File: src/main/java/com/greenfoxacademy/goddesstribesbackend/controllers/UserController.java (hunk 3 above)
Comment (ZetenyiEmese) [reply src/main/java/com/greenfoxacademy/goddesstribesbackend/controllers/UserController.java]: No, else is not needed. It was removed.

File: src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTAuthenticationFilter.java (hunk 4 above)
Comment (ZetenyiEmese) [reply src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTAuthenticationFilter.java]: Done.

File: src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTUtility.java (hunk 5 above)
Comment (ZetenyiEmese) [reply src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTUtility.java]: Ok, removed.

File: src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTUtility.java (hunk 6 above)
Comment (ZetenyiEmese) [reply src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTUtility.java]: Ok, removed.

File: src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTUtility.java (hunk 7 above)
Comment (ZetenyiEmese) [reply src/main/java/com/greenfoxacademy/goddesstribesbackend/security/jwt/JWTUtility.java]: I always add an empty line after the last line, but it is not visible on github...sorry

**Review by balazsbarni — APPROVED** [review]

**Review by symunona — APPROVED** [review]
The one and only thing I found, was the name of the branch :)
We will have a chat about that, it's good to go now!

**Pull request merged**